### PR TITLE
Refactor ImageRecordIter

### PR DESCRIPTION
**Pull request by ZhennanQin**
## Description ##
This PR brings below changes to ImageRecordIter:

* Add new parameter `dtype`, making ImageRecordIter(dtype='uint8') equivalent to  ImageRecordUInt8Iter.
* Add new optional parameter `ctx`, which indicates the device context used for. When `ctx='cpu'` is specified,  a CPU backend optimized data loader will be used.
* Add new CPU backend optimized implementation. In this implementation, data_loader is working as a engine operator. Overall throughput get improved, and allow profiling data_loader overhead with built-in profiler.

@pengzhao-intel @TaoLv @xinyu-intel @anirudh2290 

### Essentials ###
Please feel free to remove inapplicable items for your PR.
- [ ] The PR title starts with [MXNET-$JIRA_ID], where $JIRA_ID refers to the relevant [JIRA issue](https://issues.apache.org/jira/projects/MXNET/issues) created (except PRs with tiny changes)
- [ ] Changes are complete (i.e. I finished coding on this PR)
- [ ] All changes have test coverage:
- Unit tests are added for small changes to verify correctness (e.g. adding a new operator)
- Nightly tests are added for complicated/long-running ones (e.g. changing distributed kvstore)
- Build tests will be added for build configuration changes (e.g. adding a new build option with NCCL)
- [ ] Code is well-documented: 
- For user-facing API changes, API doc string has been updated. 
- For new C++ functions in header files, their functionalities and arguments are documented. 
- For new examples, README.md is added to explain the what the example does, the source of the dataset, expected performance on test set and reference to the original paper if applicable
- Check the API doc at http://mxnet-ci-doc.s3-accelerate.dualstack.amazonaws.com/PR-$PR_ID/$BUILD_ID/index.html
- [ ] To the my best knowledge, examples are either not affected by this change, or have been fixed to be compatible with this change

### Changes ###
- [ ] Feature1, tests, (and when applicable, API doc)
- [ ] Feature2, tests, (and when applicable, API doc)

## Comments ##
- If this change is a backward incompatible change, why must this change be made.
- Interesting edge cases to note here

### CODE REVIEW

File: src/io/iter_image_recordio_2.cc [hunk 1]
```diff
@@ -133,15 +134,14 @@ inline void ImageRecordIOParser2<DType>::Init(
...
-  prefetch_param_.InitAllowUnknown(kwargs);
   n_parsed_ = 0;
   overflow = false;
   rnd_.seed(kRandMagic + record_param_.seed);
   int maxthread, threadget;
   #pragma omp parallel
   {
     // be conservative, set number of real cores
-    maxthread = std::max(omp_get_num_procs() / 2 - 1, 1);
+    maxthread = std::max(omp_get_num_procs(), 1);
```
Comment (pengzhao-intel) [inline src/io/iter_image_recordio_2.cc]: Does this get the # of logic cores?

Comment (ZhennanQin) [reply src/io/iter_image_recordio_2.cc]: Yes. I think we should let user to make the decision, just as normal operator.  Before this change, data loader will only use n/2 cores for `export OMP_NUM_THREADS=n`.

Comment (pengzhao-intel) [reply src/io/iter_image_recordio_2.cc]: Make sense

**Comment by pengzhao-intel** [comment]
@wkcn could you help take a review?

**Comment by pengzhao-intel** [comment]
@szha to confirm the API enhancement.

**Comment by roywei** [comment]
@mxnet-label-bot add [Data-loading]

**Comment by pengzhao-intel** [comment]
@xinyu-intel please help rebase the code and paste the performance data as the request from @wkcn

File: src/io/iter_image_recordio_2.cc (hunk 1 above)
Comment (zhreshold) [reply src/io/iter_image_recordio_2.cc]: Since now the iterator is pushed to engine, will this omp threading make troubles?

**Pull request merged**